Individual Shipment Updater (Redux)

DanielWright opened the pull request:
This PR proposes to extract an individual shipment updater object from the existing ShipmentsUpdater, so that the IndividualShipmentUpdater can be overridden or extended by specific integrations.

A quasi-reimplementation of #14.

Commit:
Minor Tooling Tweaks, Delinting

- Enable focus filtering in RSpec
- Use relative paths in .gitignore
- Whitespace linting

Commit:
Scaffolds Extraction of IndividualShipmentUpdater

It would be nice to extract the behaviour in ShipmentsUpdater that updates each
individual shipment, so that specific integrations could augment its behaviour
with less impact. This commit adds a set of pending specs that describe the
behaviour of ShipmentsUpdater#update_shipment such that it can be refactored
into IndividualShipmentUpdater, and then ultimately refactored further.

Commit:
Implementation of IndividualShipmentUpdater

The behaviours in ShipmentsUpdater#update_shipment have been extracted into a
separate object, which can be more easily extended by host apps.

Commit:
Configurable Individual Shipment Updater

Host applications can now supply an individual-shipment updater in the
ShipFosdick configuration, if desired.

🎩 @peterberkenbosch

DanielWright commented on lib/ship_fosdick/individual_shipment_updater.rb:
```diff
@@ -0,0 +1,68 @@
+    attr_reader :invalid_transition_error_factory, :manifest_row,
+      :shipment_repository
+
+    def can_update?
```
Worth extracting this to a policy object? Probably not yet.

hugobast replied:
I agree with not yet.

DanielWright commented on lib/ship_fosdick/individual_shipment_updater.rb:
```diff
@@ -0,0 +1,68 @@
+      shipment.order.updater.update
+    end
+
+    class ManifestRow < Struct.new(:row)
```
As an internal class, how can this be extended? Would it be better to complete the abstraction by providing an interface to each field in the row, or to promote the value-object so it can be extended?

DanielWright commented on lib/ship_fosdick/individual_shipment_updater.rb:
```diff
@@ -0,0 +1,68 @@
+    end
+
+    def update_order
+      shipment.order.updater.update_shipment_state
```
Can this be replaced with a simple [`order.update!`][1] call? It calls `updater.update`, and [`updater.update`][2] calls `update_shipment_state`.

[1]: https://github.com/spree/spree/blob/2-4-stable/core/app/models/spree/order.rb#L249-L251
[2]: https://github.com/spree/spree/blob/2-4-stable/core/app/models/spree/order_updater.rb#L17-L26

hugobast replied:
That's a lot of chaining action, so I agree, is this a painful move?

hugobast replied:
I'm not seeing where `update_shipment_state` would be defined, is it missing?

DanielWright replied:
It's in the [`OrderUpdater`][1].

[1]: https://github.com/spree/spree/blob/2-4-stable/core/app/models/spree/order_updater.rb#L125-L143

Commit:
Adds Unit Specs for Fosdick::ManifestFile

These basic specs will provide a basis for refactoring.

Commit:
Refactors ShipFosdick::ManifestFile

- injects the manifest repository dependency
- refactors ManifestRow value-object as internal class of ManifestFile, rather
  than IndividualShipmentUpdater
- tidies up ShipmentsUpdater specs
- memoizes IndividualShipmentUpdater#shipment for improved performance
- improves integration spec to demonstrate the persisted tracking code

hugobast commented on README.md:
```diff
@@ -27,6 +27,7 @@ ShipFosdick.configure do |config|
   config.client_name = ENV.fetch('CLIENT_NAME')
   config.client_code = ENV.fetch('CLIENT_CODE')
   config.adcode = ENV.fetch('ADCODE')
+  config.shipment_updater_class = "ShipFosdick::IndividualShipmentUpdater"
```
Out of curiosity, I've made something similar for QL, but I went with a callable object (class, block, w/e). Here my question is, what's the tradeoff between referencing the class by it's constant and the string I see above?

DanielWright replied:
The main one I can think of is you can supply the classname in an environment variable. Is that a good enough reason? What do you think?

hugobast replied:
Yeah that's a cool use case, you've convinced me.

hugobast commented on lib/ship_fosdick/individual_shipment_updater.rb:
```diff
@@ -0,0 +1,68 @@
+
+      raise_invalid_transition_error unless shipment.can_ship?
+
+      update_attributes
```
I am gonna say, let's have this one raise an error, and I'm wondering now if `update_order` should also raise?

DanielWright replied:
I feel like this opens a real can of worms, though. Then this object has to raise all _kinds_ of errors: an error when the shipment can't be found, an error when the shipment is in the wrong state, etc. How do you think that should be handled?

hugobast replied:
Wish I had a good answer, could we log the error (any error) instead?

Commit:
Refactors IndividualShipmentUpdater

Replaces two train-wrecky calls, used to update the order, with a single call.
The single call still runs afoul of Demeter, necessitating a second custom
exception in the extremely unlikely case a shipment has no associated order.

Commit:
A Failed Update Raises an Exception

When the IndividualShipmentUpdater cannot update a shipment, instead of failing
silently, it will now raise a ShipFosdick::FailedUpdateError. Additionally, the
update is performed atomically, so shipments should remain in a reliable state
even on failures.

DanielWright commented:
@hugobast Let me know how the updated updater looks to you.

hugobast commented on lib/ship_fosdick/individual_shipment_updater.rb:
```diff
@@ -1,41 +1,55 @@
+
+    def atomically(&block)
+      ActiveRecord::Base.transaction(&block)
+    end
```
That's neat!

hugobast commented:
👍
